compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs

```diff
@@ -2130,21 +2130,27 @@ impl<'cx, 'tcx> MirBorrowckCtxt<'cx, 'tcx> {
                 /// misleading users in cases like `tests/ui/nll/borrowed-temporary-error.rs`.
                 /// We could expand the analysis to suggest hoising all of the relevant parts of
                 /// the users' code to make the code compile, but that could be too much.
-                struct NestedStatementVisitor {
+                /// We found the `prop_expr` by the way to check whether the expression is a `FormatArguments`,
+                /// which is a special case since it's generated by the compiler.
+                struct NestedStatementVisitor<'tcx> {
                     span: Span,
                     current: usize,
                     found: usize,
+                    prop_expr: Option<&'tcx hir::Expr<'tcx>>,
                 }
 
-                impl<'tcx> Visitor<'tcx> for NestedStatementVisitor {
-                    fn visit_block(&mut self, block: &hir::Block<'tcx>) {
+                impl<'tcx> Visitor<'tcx> for NestedStatementVisitor<'tcx> {
+                    fn visit_block(&mut self, block: &'tcx hir::Block<'tcx>) {
                         self.current += 1;
                         walk_block(self, block);
                         self.current -= 1;
                     }
-                    fn visit_expr(&mut self, expr: &hir::Expr<'tcx>) {
+                    fn visit_expr(&mut self, expr: &'tcx hir::Expr<'tcx>) {
                         if self.span == expr.span.source_callsite() {
                             self.found = self.current;
+                            if self.prop_expr.is_none() {
```

Comment:
What if we have nested `format_args` expressions?

Do we only care about the case where we assign the result of `format_args!`? If so, why can't we just check for `stmt` being an assignment where the rhs expr `is_format_args_item`?

Reply:
the `if self.span == expr.span.source_callsite()` guarantee we find out the expr we're intrested in, the code goes here means a borrow(mostly binding) happens, I think use span to find expr is mostly enough for suggestions.

Reply:
What does your branch output on something like this: `let x = format_args!("a {} {} {}.", 1, format_args!("b{}!", 2), 3)`? Can you maybe add this to your test?

Reply:
this is valid code, the output is:
```console
warning: unused variable: `x`
 --> tests/ui/borrowck/issue-114374-invalid-help-fmt-args.rs:5:9
  |
5 |     let x = format_args!("a {} {} {}.", 1, format_args!("b{}!", 2), 3);
  |         ^ help: if this is intentional, prefix it with an underscore: `_x`
  |
  = note: `#[warn(unused_variables)]` on by default
```
if we use `_x` the code will compile successfully.
This means this code will not get into the code path in this PR.
I added it to the test case.